airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py

```diff
@@ -52,14 +58,54 @@ def s3_client(self) -> BaseClient:
             raise ValueError("Source config is missing; cannot create the S3 client.")
         if self._s3_client is None:
             client_kv_args = _get_s3_compatible_client_args(self.config) if self.config.endpoint else {}
-            self._s3_client = boto3.client(
-                "s3",
-                aws_access_key_id=self.config.aws_access_key_id,
-                aws_secret_access_key=self.config.aws_secret_access_key,
-                **client_kv_args,
-            )
+
+            if self.config.role_arn:
+                self._s3_client = self._get_iam_s3_client(client_kv_args)
+            else:
+                self._s3_client = boto3.client(
+                    "s3",
+                    aws_access_key_id=self.config.aws_access_key_id,
+                    aws_secret_access_key=self.config.aws_secret_access_key,
+                    **client_kv_args,
+                )
+
         return self._s3_client
 
+    def _get_iam_s3_client(self, client_kv_args: dict) -> BaseClient:
```

Comment:
This is a nice solution. Can you link any documentation you used for this? I found this article that gave a nice description: https://medium.com/@li.chastina/auto-refresh-aws-tokens-using-iam-role-and-boto3-afd3c52fd8c7

Reply:
I also used the same Medium article but didn't find any additional official documentation to link except for the source code, so I will not add any links.

Reply:
Can you add docstring for this method.

Reply:
Added docstring